## Lab : Medical Insurance Data
### Name : Mohamed Abdalkader Abdalsalam

Our problem we have a medical insurance dataset that descripe where people are always confused about their medical insurance and don't know the cost of insurance at different ages and conditions. This data is useful for these people and is useful to make predictions of the insurance cost they will have to pay

<font size = 6>Libraries</font><br>
<font size = 2 color = 'gray'>just run that cell</font>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


here we will explore our data to ensure that our data is clean
to get use of data mining techniques prevent any data mining algorithm to get biased, catch unneeded patterns
**Garbig in Garbig out** 

In [2]:
data = pd.read_csv("Medical Insurance/Train_Data.csv")
data.head()


age     sex        bmi smoker     region  children       charges
0  21.000000    male  25.745000     no  northeast         2   3279.868550
1  36.976978  female  25.744165    yes  southeast         3  21454.494239
2  18.000000    male  30.030000     no  southeast         1   1720.353700
3  37.000000    male  30.676891     no  northeast         3   6801.437542
4  58.000000    male  32.010000     no  southeast         1  11946.625900

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3630 entries, 0 to 3629
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       3630 non-null   float64
 1   sex       3630 non-null   object 
 2   bmi       3630 non-null   float64
 3   smoker    3630 non-null   object 
 4   region    3630 non-null   object 
 5   children  3630 non-null   int64  
 6   charges   3630 non-null   float64
dtypes: float64(3), int64(1), object(3)
memory usage: 198.6+ KB


**Q:** Write a Python code that count null values in each attributes.

In [4]:
data.isna().sum()

age         0
sex         0
bmi         0
smoker      0
region      0
children    0
charges     0
dtype: int64

from data :<br>
    1- Is there is any Null Values? If any in which attributes ? and How many?<br>
    2- What is data type of each attributes ?<br>
<font color='blue'>Answer</font><br>
There are no nulls

**Q:** Write a Python code that count Duplicates rows/records in the data.

In [5]:
data.duplicated().sum()

679

Is there is any Duplicates? If any How many?<br>
<font color='blue'>Answer</font><br> 
Yes, There are 679 dublicated rows

**Q:** Write a Python code that handle Duplicates rows/records in the data.

In [6]:
data.drop_duplicates(inplace=True)
data

age     sex        bmi smoker     region  children       charges
0     21.000000    male  25.745000     no  northeast         2   3279.868550
1     36.976978  female  25.744165    yes  southeast         3  21454.494239
2     18.000000    male  30.030000     no  southeast         1   1720.353700
3     37.000000    male  30.676891     no  northeast         3   6801.437542
4     58.000000    male  32.010000     no  southeast         1  11946.625900
...         ...     ...        ...    ...        ...       ...           ...
2947  43.000000    male  35.531966    yes  southeast         2  18989.365540
2948  31.374148    male  28.278549     no  northeast         4  21806.181609
2949  50.370750  female  26.766297     no  northwest         4  11355.817586
2950  49.201678  female  31.396840     no  northwest         4  11185.914056
2951  49.000000    male  30.900000    yes  southwest         0  39727.614000

[2951 rows x 7 columns]

Is it fine to loose some data?<br>
yes in our case as it's just redundent data **No new information**  

### 

In [7]:
attributes = data.drop("charges", axis=1)
label = data['charges']

In [8]:
attributes.describe()

age          bmi     children
count  2951.000000  2951.000000  2951.000000
mean     38.972994    30.614864     2.498814
std      12.191312     5.412875     1.711947
min      18.000000    15.960000     0.000000
25%      29.000000    26.717292     1.000000
50%      39.296556    30.210000     3.000000
75%      48.559226    34.095907     4.000000
max      64.000000    53.130000     5.000000

1- what are attributes result from describe function?<br>
2- Name attributes that are missing and reason.<br>
<font color='blue'>Answer</font><br>

- what are attributes result from describe function?
    - age, bmi, children
- Name attributes that are missing and reason
    - sex, smoker, region are categorical columns
    - charges was dropped before using describe function

1- what is/are **Nominal** attributes in that dataset?<br>
2- what is/are **Binary** attributes in that dataset?<br>
3- what is/are **Symmetric Binary** attributes in that dataset?<br>
4- what is/are **Asymmetric Binary** attributes in that dataset?<br>
5- what is/are **Continuous** attributes in that dataset?<br>
6- what is/are **Discrete** attributes in that dataset?<br>
<font color='blue'>Answer</font><br>
- region
- sex and smoker
- sex and smoker
- None
- age, bmi and children
- children

### <font color = 'Blue'>Preprocessing Features</font>

in that part we will prepare or handling features/attributes accourding to it's datatype
first we have to encode(convert) any string to numerical value{**sex** , **smoker** , **region**}
but each attribute will have it own encoding
second we will handle continuous features {**Age**,**bmi**} for **Age** we need to round it but **bmi** it's just fine
finally discrete features needs no handling just left as it is

#### Sex

In [9]:
# sex

print("Unique Values in Sex Attribute :", data['sex'].unique())

Unique Values in Sex Attribute : ['male' 'female']


**Q:** Write a Python code to convert **Sex** attribute where male to 0 and female to 1 using sex_wrapper dictionary .

In [10]:
sex_wrapper = {'male':0, 'female':1}
data.sex = data.sex.replace(sex_wrapper)
data.head()

age  sex        bmi smoker     region  children       charges
0  21.000000    0  25.745000     no  northeast         2   3279.868550
1  36.976978    1  25.744165    yes  southeast         3  21454.494239
2  18.000000    0  30.030000     no  southeast         1   1720.353700
3  37.000000    0  30.676891     no  northeast         3   6801.437542
4  58.000000    0  32.010000     no  southeast         1  11946.625900

#### Smoker

In [11]:
# smoker
print("Unique Values in Smoker Attribute :", attributes['smoker'].unique())

Unique Values in Smoker Attribute : ['no' 'yes']


**Q:** Write a Python code to convert **Smoker** attribute where No to 0 and Yes to 1 using smoker_wrapper dictionary .

In [12]:
smoker_wrapper = {'no':0, 'yes':1}
data.smoker = data.smoker.replace(smoker_wrapper)
data.head()

age  sex        bmi  smoker     region  children       charges
0  21.000000    0  25.745000       0  northeast         2   3279.868550
1  36.976978    1  25.744165       1  southeast         3  21454.494239
2  18.000000    0  30.030000       0  southeast         1   1720.353700
3  37.000000    0  30.676891       0  northeast         3   6801.437542
4  58.000000    0  32.010000       0  southeast         1  11946.625900

#### Region

In [13]:
# region 
print("Unique Values in Region Attribute :", attributes['region'].unique())

Unique Values in Region Attribute : ['northeast' 'southeast' 'northwest' 'southwest']


As we saw we have distinct 4 values for Region attribute so How we will handle it ?
One-hot encodding what is that? <br>
simply it creating new 4 columns/features/attributes named after unique values and all these features carries either 0 (if it's NOT the value) 1 (if it carries that value)

**Q:** Write a Python code to One Hot Encode **Region** attribute .

In [14]:
data = pd.get_dummies(data, columns=['region'])
data

age  sex        bmi  smoker  children       charges  \
0     21.000000    0  25.745000       0         2   3279.868550   
1     36.976978    1  25.744165       1         3  21454.494239   
2     18.000000    0  30.030000       0         1   1720.353700   
3     37.000000    0  30.676891       0         3   6801.437542   
4     58.000000    0  32.010000       0         1  11946.625900   
...         ...  ...        ...     ...       ...           ...   
2947  43.000000    0  35.531966       1         2  18989.365540   
2948  31.374148    0  28.278549       0         4  21806.181609   
2949  50.370750    1  26.766297       0         4  11355.817586   
2950  49.201678    1  31.396840       0         4  11185.914056   
2951  49.000000    0  30.900000       1         0  39727.614000   

      region_northeast  region_northwest  region_southeast  region_southwest  
0                    1                 0                 0                 0  
1                    0                 0                 1                 0  
2                    0                 0                 1                 0  
3                    1                 0                 0                 0  
4                    0                 0                 1                 0  
...                ...               ...               ...               ...  
2947                 0                 0                 1                 0  
2948                 1                 0                 0                 0  
2949                 0                 1                 0                 0  
2950                 0                 1                 0                 0  
2951                 0                 0                 0                 1  

[2951 rows x 10 columns]

<font color = 'blue'>Note</font> : we can use get_dummies function instead of using dictionary, I just want you practice on different implementation

#### Age

**Q:** Write a Python code to round up and down **Age** attribute .

In [15]:
# age

data.age = round(data.age)
data

age  sex        bmi  smoker  children       charges  region_northeast  \
0     21.0    0  25.745000       0         2   3279.868550                 1   
1     37.0    1  25.744165       1         3  21454.494239                 0   
2     18.0    0  30.030000       0         1   1720.353700                 0   
3     37.0    0  30.676891       0         3   6801.437542                 1   
4     58.0    0  32.010000       0         1  11946.625900                 0   
...    ...  ...        ...     ...       ...           ...               ...   
2947  43.0    0  35.531966       1         2  18989.365540                 0   
2948  31.0    0  28.278549       0         4  21806.181609                 1   
2949  50.0    1  26.766297       0         4  11355.817586                 0   
2950  49.0    1  31.396840       0         4  11185.914056                 0   
2951  49.0    0  30.900000       1         0  39727.614000                 0   

      region_northwest  region_southeast  region_southwest  
0                    0                 0                 0  
1                    0                 1                 0  
2                    0                 1                 0  
3                    0                 0                 0  
4                    0                 1                 0  
...                ...               ...               ...  
2947                 0                 1                 0  
2948                 0                 0                 0  
2949                 1                 0                 0  
2950                 1                 0                 0  
2951                 0                 0                 1  

[2951 rows x 10 columns]

Final Data "The Cleaned one"

In [16]:
data.describe()

age          sex          bmi       smoker     children  \
count  2951.000000  2951.000000  2951.000000  2951.000000  2951.000000   
mean     38.973568     0.442223    30.614864     0.156218     2.498814   
std      12.192272     0.496735     5.412875     0.363124     1.711947   
min      18.000000     0.000000    15.960000     0.000000     0.000000   
25%      29.000000     0.000000    26.717292     0.000000     1.000000   
50%      39.000000     0.000000    30.210000     0.000000     3.000000   
75%      49.000000     1.000000    34.095907     0.000000     4.000000   
max      64.000000     1.000000    53.130000     1.000000     5.000000   

            charges  region_northeast  region_northwest  region_southeast  \
count   2951.000000       2951.000000       2951.000000       2951.000000   
mean   12847.623458          0.236869          0.248729          0.279227   
std    10762.255653          0.425233          0.432350          0.448696   
min     1121.873900          0.000000          0.000000          0.000000   
25%     5701.034262          0.000000          0.000000          0.000000   
50%     9523.978586          0.000000          0.000000          0.000000   
75%    14831.372859          0.000000          0.000000          1.000000   
max    63770.428010          1.000000          1.000000          1.000000   

       region_southwest  
count       2951.000000  
mean           0.235175  
std            0.424180  
min            0.000000  
25%            0.000000  
50%            0.000000  
75%            0.000000  
max            1.000000

<center> <font size = 6 color = 'crimson'>Finally Finished The First Part  !!!</font>

In [17]:
# data.to_csv("HandledMedicalInsuranceData.csv")